src/ofonotracker.cpp

```diff
@@ -0,0 +1,335 @@
+        watcher->deleteLater();
+    });
+}
+
+void OFonoTrackerPrivate::getModemsSync()
+{
+    QDBusMessage request(QDBusMessage::createMethodCall(OFONO_SERVICE, OFONO_MANAGER_OBJECT, OFONO_MANAGER_INTERFACE, OFONO_MANAGER_GET_MODEMS));
+    request.setAutoStartService(false);
+    QDBusReply<ModemInfoList> response(m_systemBus.call(request));
+    if (response.isValid())
+        setModems(response.value());
+    else if (response.error().name() != DBUS_ERROR_NAME_HAS_NO_OWNER)
+        qWarning() << response.error();
+    evaluateProperties();
+}
+
+void OFonoTrackerPrivate::flushModems()
+{
+    if (!m_modemInfoHash.isEmpty()) {
+        m_modemInfoHash.clear();
+        m_modemPathList.clear();
+        evaluatePropertiesLater();
+    }
+}
+
+void OFonoTrackerPrivate::setModems(const ModemInfoList &modemInfoList)
+{
+    flushModems();
+    for (auto iter = modemInfoList.cbegin(); iter != modemInfoList.cend(); ++iter)
+        addModem(*iter);
+}
+
+void OFonoTrackerPrivate::addModem(const ModemInfo &modemInfo)
+{
+    QString path(modemInfo.m_objectPath.path());
+    if (!m_modemInfoHash.contains(path)) {
+        qWarning() << "#### modem:" << path << "added";
+        m_modemPathList.append(path);
+        m_modemInfoHash.insert(path, ModemInfo(path));
+        evaluatePropertiesLater();
+    }
+    for (auto iter = modemInfo.m_properties.cbegin(); iter != modemInfo.m_properties.cend(); ++iter)
+        setModemProperty(path, iter.key(), iter.value());
+}
+
+void OFonoTrackerPrivate::removeModem(QString path)
+{
+    if (m_modemInfoHash.remove(path) > 0) {
+        qWarning() << "#### modem:" << path << "removed";
+        m_modemPathList.removeOne(path);
+        evaluatePropertiesLater();
+    }
+}
+
+void OFonoTrackerPrivate::setModemProperty(QString path, QString key, QVariant val)
+{
+    if (trackedModemProperty(key) && m_modemInfoHash.contains(path)) {
+        QVariant &cached(m_modemInfoHash[path].m_properties[key]);
+        if (cached != val) {
+            qWarning() << "#### modem:" << path << "prop:" << key << ":=" << val;
+            cached = val;
+            evaluatePropertiesLater();
+        }
+    }
+}
+
+bool OFonoTrackerPrivate::trackedModemProperty(const QString &key)
+{
+    return key == OFOFO_MODEM_PROPERTY_SERIAL;
+}
+
+/* ========================================================================= *
+ * OFonoTracker
+ * ========================================================================= */
+
+OFonoTracker::OFonoTracker(bool synchronousInit, QObject *parent)
+    : QObject(parent)
+    , d_ptr(new OFonoTrackerPrivate(synchronousInit, this))
+{
+}
+
+OFonoTracker::~OFonoTracker()
+{
+}
+
+QStringList OFonoTracker::imeiNumbers() const
+{
+    Q_D(const OFonoTracker);
+    return d->imeiNumbers();
+}
+
+/* ========================================================================= *
+ * MOC
+ * ========================================================================= */
```

Comment:
Perhaps obvious :)

Reply:
Will remove - although nothing about moc is obvious to me ;-/